src/librbd/cache/ReplicatedWriteLog.cc

```diff
@@ -0,0 +1,1664 @@
+
+  Context *invalidate_ctx = new FunctionContext(
+    [this, offset, length, invalidate_done_ctx](int r) {
+      invalidate({{offset, length}}, invalidate_done_ctx);
+    });
+  flush(invalidate_ctx);
+  */
+}
+
+template <typename I>
+void ReplicatedWriteLog<I>::aio_flush(Context *on_finish) {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 20) << "on_finish=" << on_finish << dendl;
+
+  {
+    RWLock::RLocker snap_locker(m_image_ctx.snap_lock);
+    if (m_image_ctx.snap_id != CEPH_NOSNAP || m_image_ctx.read_only) {
+      on_finish->complete(-EROFS);
+      return;
+    }
+  }
+
+  if (!m_flush_seen) {
+    Mutex::Locker locker(m_lock);
+    if (!m_flush_seen) {
+      ldout(cct, 5) << "flush seen" << dendl;
+      m_flush_seen = true;
+      if (!m_persist_on_flush && m_persist_on_write_until_flush) {
+	m_persist_on_flush = true;
+	ldout(cct, 5) << "now persisting on flush" << dendl;
+      }
+    }
+  }
+  
+  if (m_image_ctx.persistent_cache_enabled) {
+    m_image_cache.aio_flush(on_finish);
+  } else {
+    m_image_writeback.aio_flush(on_finish);
+  }
+  /*
+  Context *ctx = new FunctionContext(
+    [this, on_finish](int r) {
+      if (r < 0) {
+        on_finish->complete(r);
+      }
+      m_image_writeback.aio_flush(on_finish);
+    });
+
+  flush(ctx);
+  */
+}
+
+template <typename I>
+void ReplicatedWriteLog<I>::aio_writesame(uint64_t offset, uint64_t length,
+                                      bufferlist&& bl, int fadvise_flags,
+                                      Context *on_finish) {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 20) << "offset=" << offset << ", "
+                 << "length=" << length << ", "
+                 << "data_len=" << bl.length() << ", "
+                 << "on_finish=" << on_finish << dendl;
+  {
+    RWLock::RLocker snap_locker(m_image_ctx.snap_lock);
+    if (m_image_ctx.snap_id != CEPH_NOSNAP || m_image_ctx.read_only) {
+      on_finish->complete(-EROFS);
+      return;
+    }
+  }
+
+  if (m_image_ctx.persistent_cache_enabled) {
+    m_image_cache.aio_writesame(offset, length, std::move(bl), fadvise_flags, on_finish);
+  } else {
+    m_image_writeback.aio_writesame(offset, length, std::move(bl), fadvise_flags, on_finish);
+  }
+
+  /*
+  bufferlist total_bl;
+
+  uint64_t left = length;
+  while(left) {
+    total_bl.append(bl);
+    left -= bl.length();
+  }
+  assert(length == total_bl.length());
+  aio_write({{offset, length}}, std::move(total_bl), fadvise_flags, on_finish);
+  */
+}
+
+template <typename I>
+void ReplicatedWriteLog<I>::aio_compare_and_write(Extents &&image_extents,
+                                                     bufferlist&& cmp_bl,
+                                                     bufferlist&& bl,
+                                                     uint64_t *mismatch_offset,
+                                                     int fadvise_flags,
+                                                     Context *on_finish) {
+
+  // TODO:
```

Comment:
The real use-case for this is for atomic test and set SCSI operations on shared images, therefore the cache shouldn't be in-use. Worst case you could throw an error to indicate it's not supported -- otherwise block guard it, flush if blocks are currently cached, and pass-through the request.

Reply:
Right, the blockguard will take care of this. It's not super complicated, but you need a bunch of testing to verify it's working, and it's really hitting cache when it says it works. There's a bunch of general issues like that about the unit tests forcing some internal states in the cache (pause or force flushing to ensure hits or misses occur, etc.) that I need to figure out. It seems like some private interfaces for use by the tests, but something about that feels wrong. There must be a better way.

Unmap and write same are more work. We probably don't want to wait for them to write through, but we can't transform them into bare writes. The log entries will need some flags to indicate they were write same's and/or unmaps so the thin provisioning happens.